File: python3/26.remove-duplicates-from-sorted-array.py
#
# @lc app=leetcode id=26 lang=python3
#
# [26] Remove Duplicates from Sorted Array
#
# Time: O(n)
# Space: O(1)

# @lc code=start
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        uniq = 0
        for currNum in nums:
            if nums[uniq] != currNum:
                uniq += 1
                nums[uniq] = currNum
        return uniq + 1
# @lc code=end

File: C++/26_Remove_Duplicates_From_Sorted_Array.cpp
/***                                                *
 * 1. Two Pointers                                  *                              
 * Time complexity: O(n)                            *
 * Space complexity: O(1)                         ***/

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        int index = 0;
        for(int i = 0; i < nums.size(); i++) {
            if(index < 1 || nums[i] > nums[index - 1]) {
                nums[index++] = nums[i];
            }
        }
        
        return index;
    }
};

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        
        if(nums.size() == 0) {
            return 0;
        }
        
        int index = 0;
        for(int i = 1; i < nums.size(); i++) {
            if(nums[i] != nums[index]) {
                nums[++index] = nums[i];
            }
        }
        
        return index + 1;
    }
};